File: go/funding_raised.go
package fundingraised

import (
	"bufio"
	"encoding/csv"
	"errors"
	"io"
	"os"
)

func Where(options map[string]string) []map[string]string {
	f, _ := os.Open("../startup_funding.csv")
	reader := csv.NewReader(bufio.NewReader(f))
	csv_data := [][]string{}

	for {
		row, err := reader.Read()

		if err == io.EOF {
			break
		}

		csv_data = append(csv_data, row)
	}

	_, ok := options["company_name"]
	if ok == true {
		results := [][]string{}
		for i := 0; i < len(csv_data); i++ {
			if csv_data[i][1] == options["company_name"] {
				results = append(results, csv_data[i])
			}
		}
		csv_data = results
	}

	_, ok = options["city"]
	if ok == true {
		results := [][]string{}
		for i := 0; i < len(csv_data); i++ {
			if csv_data[i][4] == options["city"] {
				results = append(results, csv_data[i])
			}
		}
		csv_data = results
	}

	_, ok = options["state"]
	if ok == true {
		results := [][]string{}
		for i := 0; i < len(csv_data); i++ {
			if csv_data[i][5] == options["state"] {
				results = append(results, csv_data[i])
			}
		}
		csv_data = results
	}

	_, ok = options["round"]
	if ok == true {
		results := [][]string{}
		for i := 0; i < len(csv_data); i++ {
			if csv_data[i][9] == options["round"] {
				results = append(results, csv_data[i])
			}
		}
		csv_data = results
	}

	output := []map[string]string{}
	for i := 0; i < len(csv_data); i++ {
		mapped := make(map[string]string)
		mapped["permalink"] = csv_data[i][0]
		mapped["company_name"] = csv_data[i][1]
		mapped["number_employees"] = csv_data[i][2]
		mapped["category"] = csv_data[i][3]
		mapped["city"] = csv_data[i][4]
		mapped["state"] = csv_data[i][5]
		mapped["funded_date"] = csv_data[i][6]
		mapped["raised_amount"] = csv_data[i][7]
		mapped["raised_currency"] = csv_data[i][8]
		mapped["round"] = csv_data[i][9]
		output = append(output, mapped)
	}

	return output
}

func FindBy(options map[string]string) (map[string]string, error) {
	f, _ := os.Open("../startup_funding.csv")
	reader := csv.NewReader(bufio.NewReader(f))
	csv_data := [][]string{}

	for {
		row, err := reader.Read()

		if err == io.EOF {
			break
		}

		csv_data = append(csv_data, row)
	}

	for i := 0; i < len(csv_data); i++ {
		var ok bool
		mapped := make(map[string]string)

		_, ok = options["company_name"]
		if ok == true {
			if csv_data[i][1] == options["company_name"] {
				mapped["permalink"] = csv_data[i][0]
				mapped["company_name"] = csv_data[i][1]
				mapped["number_employees"] = csv_data[i][2]
				mapped["category"] = csv_data[i][3]
				mapped["city"] = csv_data[i][4]
				mapped["state"] = csv_data[i][5]
				mapped["funded_date"] = csv_data[i][6]
				mapped["raised_amount"] = csv_data[i][7]
				mapped["raised_currency"] = csv_data[i][8]
				mapped["round"] = csv_data[i][9]
			} else {
				continue
			}
		}

		_, ok = options["city"]
		if ok == true {
			if csv_data[i][4] == options["city"] {
				mapped["permalink"] = csv_data[i][0]
				mapped["company_name"] = csv_data[i][1]
				mapped["number_employees"] = csv_data[i][2]
				mapped["category"] = csv_data[i][3]
				mapped["city"] = csv_data[i][4]
				mapped["state"] = csv_data[i][5]
				mapped["funded_date"] = csv_data[i][6]
				mapped["raised_amount"] = csv_data[i][7]
				mapped["raised_currency"] = csv_data[i][8]
				mapped["round"] = csv_data[i][9]
			} else {
				continue
			}
		}

		_, ok = options["state"]
		if ok == true {
			if csv_data[i][5] == options["state"] {
				mapped["permalink"] = csv_data[i][0]
				mapped["company_name"] = csv_data[i][1]
				mapped["number_employees"] = csv_data[i][2]
				mapped["category"] = csv_data[i][3]
				mapped["city"] = csv_data[i][4]
				mapped["state"] = csv_data[i][5]
				mapped["funded_date"] = csv_data[i][6]
				mapped["raised_amount"] = csv_data[i][7]
				mapped["raised_currency"] = csv_data[i][8]
				mapped["round"] = csv_data[i][9]
			} else {
				continue
			}
		}

		_, ok = options["round"]
		if ok == true {
			if csv_data[i][9] == options["round"] {
				mapped["permalink"] = csv_data[i][0]
				mapped["company_name"] = csv_data[i][1]
				mapped["number_employees"] = csv_data[i][2]
				mapped["category"] = csv_data[i][3]
				mapped["city"] = csv_data[i][4]
				mapped["state"] = csv_data[i][5]
				mapped["funded_date"] = csv_data[i][6]
				mapped["raised_amount"] = csv_data[i][7]
				mapped["raised_currency"] = csv_data[i][8]
				mapped["round"] = csv_data[i][9]
			} else {
				continue
			}
		}

		return mapped, nil
	}

	return make(map[string]string), errors.New("Record Not Found")
}

File: ruby/funding_raised.rb
require 'csv'
require 'rspec'
require 'rspec/expectations'
require 'pry'

class FundingRaised
  def self.where(options = {})
    csv_data = CSV.read("../startup_funding.csv")

    @funding = []
    if options[:company_name]
      csv_data.select! do |row|
        row[1] == options.fetch(:company_name)
      end
    end

    if options[:city]
      csv_data.select! do |row|
        row[4] == options.fetch(:city)
      end
    end

    if options[:state]
      csv_data.select! do |row|
        row[5] == options.fetch(:state)
      end
    end

    if options[:round]
      csv_data.select! do |row|
        row[9] == options.fetch(:round)
      end
    end

    @output = []
    csv_data.each do |row|
      mapped = {}
      mapped['permalink'] = row[0]
      mapped['company_name'] = row[1]
      mapped['number_employees'] = row[2]
      mapped['category'] = row[3]
      mapped['city'] = row[4]
      mapped['state'] = row[5]
      mapped['funded_date'] = row[6]
      mapped['raised_amount'] = row[7]
      mapped['raised_currency'] = row[8]
      mapped['round'] = row[9]
      @output << mapped
    end

    @output
  end

  def self.find_by(options)
    csv_data = CSV.read("../startup_funding.csv")

    if options[:company_name]
      csv_data.each do |row|
        if row[1] == options.fetch(:company_name)
          mapped = {}
          mapped['permalink'] = row[0]
          mapped['company_name'] = row[1]
          mapped['number_employees'] = row[2]
          mapped['category'] = row[3]
          mapped['city'] = row[4]
          mapped['state'] = row[5]
          mapped['funded_date'] = row[6]
          mapped['raised_amount'] = row[7]
          mapped['raised_currency'] = row[8]
          mapped['round'] = row[9]
          return mapped
        end
      end
    end

    if options[:city]
      csv_data.each do |row|
        if row[4] == options.fetch(:city)
          mapped = {}
          mapped['permalink'] = row[0]
          mapped['company_name'] = row[1]
          mapped['number_employees'] = row[2]
          mapped['category'] = row[3]
          mapped['city'] = row[4]
          mapped['state'] = row[5]
          mapped['funded_date'] = row[6]
          mapped['raised_amount'] = row[7]
          mapped['raised_currency'] = row[8]
          mapped['round'] = row[9]
          return mapped
        end
      end
    end

    if options[:state]
      csv_data.each do |row|
        if row[5] == options.fetch(:state)
          mapped = {}
          mapped['permalink'] = row[0]
          mapped['company_name'] = row[1]
          mapped['number_employees'] = row[2]
          mapped['category'] = row[3]
          mapped['city'] = row[4]
          mapped['state'] = row[5]
          mapped['funded_date'] = row[6]
          mapped['raised_amount'] = row[7]
          mapped['raised_currency'] = row[8]
          mapped['round'] = row[9]
          return mapped
        end
      end
    end

    if options[:round]
      csv_data.each do |row|
        if row[9] == options.fetch(:round)
          mapped = {}
          mapped['permalink'] = row[0]
          mapped['company_name'] = row[1]
          mapped['number_employees'] = row[2]
          mapped['category'] = row[3]
          mapped['city'] = row[4]
          mapped['state'] = row[5]
          mapped['funded_date'] = row[6]
          mapped['raised_amount'] = row[7]
          mapped['raised_currency'] = row[8]
          mapped['round'] = row[9]
          return mapped
        end
      end
    end

    raise RecordNotFound
  end
end

class RecordNotFound < StandardError
end
